### DISCUSSION TITLE
Installation Keba Wallbox mit Modbus Fehler

### DISCUSSION THREAD

**Discussion by SpeGo23** [camueller/SmartApplianceEnabler#590] [Installation und Betrieb]
Hallo zusammen,
ich habe den SAE auf einem Raspi installiert und möchte damit meine Keba P30 PV Edition Wallbox steuern (das ist eine Variante der x-Version). Nun läuft SAE und ich kann die Wallbox per Modbus auch händisch mit SimplyModbusTCP Client lesen, das SAE Log zeigt aber error beim Lesen an (ich habe mit Register 1000 händisch im SimplyModbusTCPClient erfolgreich getestet und im SAE Log nach 1000 gefiltert, dort kommt ein Lesefehler, es kommt auch „EVModbusControl.isInErrorState“). Woran könnte es liegen, macht es Sinn wenn ich Auszüge aus dem Log hier zur Verfügung stelle? Ich habe bei der modbus Konfiguration versucht die IP Adresse der Wallbox anzugeben (irgendwo muss die doch hin?). Ansonsten habe ich strikt die Anleitung befolgt, für die Keba ist ja eigentlich schon alles in der Vorlage enthalten.
Vielen Dank im Voraus fürs Anschauen!

**Comment by camueller** [comment]
Zuerst wäre zu klären, ob der SAE überhaupt mit der Wallbox kommunizieren kann. Wenn das gegeben ist, muss man schauen, bei welchem Befehl Fehler auftreten. Möglicherweise hat Keba irgendwelche Register geändert. In jedem Fall kann man ohne Log eigentlich nichts sagen. Lade doch das Log irgendwo hoch (Google Drive Dropbox, ...) und poste einen Link, dann schaue ich es mir an.

**Comment by SpeGo23** [comment]
[rolling-2024-02-17.log](https://github.com/camueller/SmartApplianceEnabler/files/14319641/rolling-2024-02-17.log)
Danke dass du es dir anschaust! Ist es als Anhang hier auch okay? Dann muss ich nicht meine alte Dropbox reaktivieren ;-). Das Log habe ich gerade gezogen, vorhin hatte ich auch etwas mit manuellem An- und Ausschalten des Ladens über die Keba App gespielt, das ist jetzt nicht mehr im Log. Sag Bescheid, falls du dazu noch etwas brauchst.

**Comment by camueller** [comment]
Die Verbindung zur Wallbox schlägt komplett fehl:
```
java.net.ConnectException: Connection refused (Connection refused)
```

Du hast 2 Modbus konfiguriert: der mit dem Namen "modbus" zeigt auf den Raspberry Pi selbst (also `mbusd`) und der mit dem Namen "192.168.10.94" zeigt auf selbige IP-Adresse. Letzterer wird für die Verbindung zu Deiner Wallbox verwendet. Allerdings scheint sie nicht unter dieser IP-Adresse erreichbar zu sein.

**Reply by camueller** [reply]
Du hast irgendwie die IP-Adresse nicht eingegeben oder an der falschen Stelle - deshalb steht im Log
```
Connecting to modbus 192.168.10.94@:502
```
Normalerweise müsse die IP-Adresse hinter dem "@"-Zeichen stehen. Bei mir sieht die Zeile im Log entsprechend so aus:
```
2024-02-17 15:40:55,322 DEBUG [Timer-0] d.a.s.m.ModbusSlave [ModbusSlave.java:77] F-00000001-000000000019-00: Connecting to modbus wallbox@wallbox:502
2024-02-17 15:41:04,804 DEBUG [Timer-0] d.a.s.m.ModbusSlave [ModbusSlave.java:77] F-00000001-000000000019-00: Connecting to modbus modbus@127.0.0.1:502
```
Post mal einen Screenshot der Einstellungen für Modbus im SAE.

**Comment by SpeGo23** [comment]
Ich gebe zu ich habe mit den Einstellungen zuletzt etwas herumgespielt, und den modbus daher nach der IP Adresse benannt. So ist es aktuell im Setting:
![grafik](https://github.com/camueller/SmartApplianceEnabler/assets/160337201/c4d4ac69-dbc8-47d1-994b-7fa79f7cfd20)
Den Modbus "modbus" habe ich nur behalten bzw. nach dem Löschen doch nochmal hinzugefügt, weil es in den Beispielen immer da war. Der macht aber vermutlich nichts. Der andere ist nach der IP der Wallbox mittlerweile benannt und hieß ursprünglich mal "wallbox", ich hatte bei host dann die IP Adresse eingegeben. Zuletzt hatte ich auch einfach bei ID und bei Host die IP Adresse stehen, weil ich auch vermutet habe das die irgendwie fehlt. 
Sorry das aktuelle Setting ist natürlich ungeschickt, aber vielleicht hilft es trotzdem.

**Comment by SpeGo23** [comment]
Ich habe es jetzt nochmal hierhin zurück geändert:
![grafik](https://github.com/camueller/SmartApplianceEnabler/assets/160337201/2ff4ace4-d0ea-4ca7-92ea-8eed9c4bc276)
Ich könnte ein paar Minuten warten und dann nochmal das neue zugehörige Logfile schicken.

**Comment by camueller** [comment]
Mit der Einstellung sollte es passen.

**Comment by camueller** [comment]
Du musst aber auch in der Wallbox-Konfiguration den Modbus "wallbox" auswählen, nachdem Du ja die ID geändert hast!

**Comment by SpeGo23** [comment]
Danke nochmal für den Hinweis "wallbox" neu auswählen zu müssen, das Thema hatte ich vorhin auch und hätte ich jetzt trotzdem wieder vergessen. Hier nun ein neues Log, die letzten Minuten sind mit dem Setting entsprechend des Screenshots oben.
[rolling-2024-02-17_modbus_wallbox.log](https://github.com/camueller/SmartApplianceEnabler/files/14319825/rolling-2024-02-17_modbus_wallbox.log)

**Comment by camueller** [comment]
Aus dem neuen Log ist ersichtlich, dass der SAE mit der Wallbox kommuniziert, aber die Register scheinen nicht zu passen.
In der Konfiguration für die Keba gibt es kein Register 1036 - trotzdem versucht der SAE dieses Register zu lesen. Es sieht so aus, als ob die Konfiguration der Register im SAE geändert wurde.

Aber auch die Register 1000 und 1004 kann er nicht lesen, obwohl sie zur Konfiguration der Keba-Wallbox im SAE gehören. Scheinbar verwendet Deine Wallbox doch andere Register oder Keba hat diese geändert. Ich habe mal eine Blick in die Modbus-Dokoumentation Deiner Wallbox beworfen: https://www.keba.com/download/x/dea7ae6b84/kecontactp30modbustcp_pgen.pdf

Es sieht so aus, als ob Keba, auch andere Registertypen verwendet (z.B. Register 1000 scheint jetzt kein Holding-Register mehr zu sein, sondern Input-Register). Da musst Du wohl die Konfiguration aller Register prüfen und anpassen. Falls Du das nicht selbst machen kannst/willst, können wir das auch in einer [kostenpflichtigen Support-Sesstion](https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/Support_DE.md) machen.

**Comment by SpeGo23** [comment]
Danke dir für die schnelle Analyse. Ich würde mich dann selbst da mal durchgraben und falls ich es nicht hinbekomme, machen wir die Support Session. Kannst du mir vielleicht nur noch einmal Starthilfe geben? Register 1000 hatte ich mit dem Programmers Guide geprüft und dachte eigentlich, dass es immer noch so ist wie in der SAE Konfig. Wo siehst du, dass es jetzt ein Input Register ist? Sie geben doch eigentlich FC3 für die Read Register an und so ist es auch im SAE hinterlegt. Die Werte 0 und 1 sind definiert, aber in der SAE Konfig nicht hinterlegt - sonst sehe ich eigentlich nichts, was man ändern könnte.
Und Register 1036 kommt aus der Zähler Konfiguration entsprechend deines Installation Guides: https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/Keba_DE.md, ist das nicht richtig?

**Comment by camueller** [comment]
Die Vorlagen und die Dokumentation waren zu einem Zeitpunkt für bestimmte Firmware-Versionen gültig. Das kann sich geändert haben. Man muss also die aktuelle Spezifikation anschauen und die Konfiguration im SAE entsprechend anpassen. Und dann jweils schauen, ob sich die Wallbox korrekt verhält, wenn der SAE die entsprechenden Befehle schickt.
